bidscoin/dicomsort.py

```diff
@@ -70,27 +71,40 @@ def sortsession(sessionfolder: Path, dicomfiles: list, dicomfield: str, rename:
                     seriesdescr = 'unknown_protocol'
                     LOGGER.warning(f"No {dicomfield}, SeriesDecription or ProtocolName found for: {dicomfile}")
         if rename:
-            acquisitionnr = bids.get_dicomfield('AcquisitionNumber', dicomfile)
-            instancenr    = bids.get_dicomfield('InstanceNumber', dicomfile)
-            if not instancenr:
-                instancenr = bids.get_dicomfield('ImageNumber', dicomfile)          # This Attribute was named Image Number in earlier versions of this Standard
-            patientname    = bids.get_dicomfield('PatientName', dicomfile)
-            if not patientname:
-                patientname = bids.get_dicomfield('PatientsName', dicomfile)        # This Attribute was/is sometimes called PatientsName?
-
+            # Parse the naming scheme string and test if all attributes are present
+            if re.fullmatch('(({[a-zA-Z]+(:\\d+d)?})|([a-zA-Z0-9_.]+))*', rename_scheme) is None :
+                LOGGER.error('Bad naming scheme. Only alphanumeric characters could be used for the field names (with the optional number of digits afterwards, '
+                             'e.g., \'{InstanceNumber:05d}\'), and only alphanumeric characters, dots, and underscores could be used as separators. ')
+                rename = False
+            else:
+                fields = re.findall('(?<={)([a-zA-Z]+)(?::\\d+d)?(?=})', rename_scheme)
+                field_alternatives = {'InstanceNumber': 'ImageNumber', # alternative field names based on the earlier versions of the Standard or other reasons
+                                      'PatientName': 'PatientsName'}
+                bids_fields = {}
+                for field in fields:
+                    value = bids.get_dicomfield(field, dicomfile)
+                    if not value:
+                        if field in field_alternatives.keys() and bids.get_dicomfield(field_alternatives[field], dicomfile):
+                            rename_scheme.replace('{'+field+'}','{'+field_alternatives[field]+'}')
+                            bids_fields[field_alternatives[field]] = value
+
+                            LOGGER.info('{0} field is absent from the DICOM header, but {1} is present, using {1}'.format(field, field_alternatives[field]))
```

Comment:
Please use the more modern formatted string instead of using the old .format() way, it is much more readable and elegant